[tf.data] Check in a strictly faster rejection resampling transformation.

joel-shor opened the pull request:
This transformation is faster because it rejects fewer data. This
is done by occasionally sampling from the original data distribution
in an efficient way.

Tested:
bazel test :resample_test

joel-shor committed:
[tf.data] Check in a strictly faster rejection resampling
transformation.

This transformation is faster because it rejects fewer data. This
is done by occasionally sampling from the original data distribution
in an efficient way.

Tested:
bazel test :resample_test

joel-shor committed:
[tf.data] Clean up resampler and update BUILD files.

joel-shor committed:
[tf.data] Fix `absl` build rule.

joel-shor committed:
[tf.data] Reorder BUILD rule deps and add `xrange` from `six`.

joel-shor committed:
[tf.data] Second reorder BUILD rule deps.

mrry commented on tensorflow/contrib/data/python/kernel_tests/resample_test.py:
```diff
@@ -109,5 +134,55 @@ def _remap_fn(_):
+    self.assertLess(fast_time, slow_time)
+
+
+class MapDatasetBenchmark(test.Benchmark):
```
Nit: s/MapDataset/Resample/

mrry commented on tensorflow/contrib/data/python/ops/resampling.py:
```diff
@@ -50,70 +51,171 @@ def rejection_resample(class_func, target_dist, initial_dist=None, seed=None):
+
+
+def rejection_resample_v2(class_func, target_dist, initial_dist=None,
+                          seed=None):
```
Should we just replace the old version with this, and call it `rejection_resample()`?

mrry commented on tensorflow/contrib/data/python/ops/resampling.py:
```diff
@@ -50,70 +51,171 @@ def rejection_resample(class_func, target_dist, initial_dist=None, seed=None):
+
+  This differs from v1 in that it will also sample from the original dataset
+  with some probability, so it makes strictly fewer data rejections. Due to an
+  implementation detail it must initialize a separate dataset initializer, so
```
I think this is no longer true, is it?

mrry reviewed (dismissed):
New version looks good to me. Just a few minor comments and we should be good to merge.

ekelsen commented:
@joel-shor Could you address the comments and resolve the conflicts?

joel-shor committed:
[tf.data] Just replace old resample with new.

Also, add an optimization / bug fix that shortcircuits combining the two datasets if one should always be sampled from.

Tested:

bazel test :resample_test

joel-shor committed:
Merge branch 'master' into master

joel-shor commented:
Applied changes.

Note that I also added one bit of extra logic, which short-circuits the interleaving (making it faster) and avoiding a bug in one of those edge cases.

joel-shor reviewed (commented):
Done.

mrry commented on tensorflow/contrib/data/python/ops/resampling.py:
```diff
@@ -50,79 +51,184 @@ def rejection_resample(class_func, target_dist, initial_dist=None, seed=None):
+def _get_prob_original_static(initial_dist_t, target_dist_t):
+  """Returns the static probability of sampling from the original.
+
+  For some reason, `tensor_util.constant_value(prob_of_original)` of a ratio
```
> For some reason, [...].

The reason is that `tensor_util.constant_value()` only implements a handful of ops, and returns `None` when it encounters an op that it hasn't implemented. Please update the comment to make it seem less mysterious!

joel-shor replied:
Done

mrry commented on tensorflow/contrib/data/python/kernel_tests/resample_test.py:
```diff
@@ -30,52 +33,97 @@
+class ResampleTest(test.TestCase, parameterized.TestCase):
+
+  @parameterized.named_parameters(
+      ("InitialnDistributionKnown", True),
```
Typo: "Initialn" -> "Initial"

joel-shor replied:
Done

mrry reviewed (changes requested):
Just a couple of minor changes left.

joel-shor committed:
[tf.data] Make documentation changes, and add correct import.

joel-shor committed:
Merge remote-tracking branch 'origin/master'

joel-shor reviewed (commented):
Done

joel-shor commented:
Done! PTAL

On Fri, Apr 27, 2018, 2:41 AM Derek Murray <notifications@github.com> wrote:

> *@mrry* requested changes on this pull request.
>
> Just a couple of minor changes left.
> ------------------------------
>
> In tensorflow/contrib/data/python/ops/resampling.py
> <https://github.com/tensorflow/tensorflow/pull/18730#discussion_r184559610>
> :
>
> >
>    return _apply_fn
>
>
> -def _calculate_acceptance_probs(initial_probs, target_probs):
> -  """Calculate the per-class acceptance rates.
> +def _get_prob_original_static(initial_dist_t, target_dist_t):
> +  """Returns the static probability of sampling from the original.
> +
> +  For some reason, `tensor_util.constant_value(prob_of_original)` of a ratio
>
> For some reason, [...].
>
> The reason is that tensor_util.constant_value() only implements a handful
> of ops, and returns None when it encounters an op that it hasn't
> implemented. Please update the comment to make it seem less mysterious!
> ------------------------------
>
> In tensorflow/contrib/data/python/kernel_tests/resample_test.py
> <https://github.com/tensorflow/tensorflow/pull/18730#discussion_r184559789>
> :
>
> >
> -  def _testDistribution(self, initial_known):
> +  with test_obj.test_session() as sess:
> +    start_time = time.time()
> +    for _ in xrange(num_to_sample):
> +      sess.run(get_next)
> +    end_time = time.time()
> +
> +  return end_time - start_time
> +
> +
> +class ResampleTest(test.TestCase, parameterized.TestCase):
> +
> +  @parameterized.named_parameters(
> +      ("InitialnDistributionKnown", True),
>
> Typo: "Initialn" -> "Initial"
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/tensorflow/tensorflow/pull/18730#pullrequestreview-115782670>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AFvffP_gwOxIho9ESWrlP3KG9xJD2nJWks5tslsbgaJpZM4TdR7l>
> .
>

joel-shor committed:
[tf.data] Pass a Tensor to `tensor_util.constant_value` instead of possible a python list.

joel-shor committed:
[tf.data] Fix indentation.

mrry commented:
Looks like a legitimate failure on the Windows build. The usual cause for errors like that is the fact that `np.array([1]).dtype` (for any list of Python integers) is `int64` on Linux and `int32` on Windows. That probably means that some integer array created in the code needs to be explicitly cast to `int64`.

I’d start by looking at arguments to `sample_from_datasets()` although it’s possible that the bug is in the Python implementation of that function.

joel-shor committed:
[tf.data] A change to use Jenkins to test the Winsows build.

don't submit with this change!

joel-shor committed:
[tf.data] Possible bug fix to fix Winsows build.

joel-shor committed:
[tf.data] Undo previously unsuccessful bugfix, and try another one to fix the Windows build.

don't submit with this change, because it includes some debugging!

joel-shor committed:
[tf.data] Add a bunch of debugging for Jenkins to run on the Windows build.

joel-shor committed:
[tf.data] Add a bunch of debugging for Jenkins to run on the Windows build.

joel-shor committed:
[tf.data] Add a bunch of debugging for Jenkins to run on the Windows build.

joel-shor committed:
[tf.data] Fix logging ops debug statement.

joel-shor committed:
[tf.data] Properly format debug statements.

joel-shor committed:
[tf.data] Explicitly make test's dataset int64.

joel-shor committed:
[tf.data] Removed debug code.

joel-shor committed:
[tf.data] More debug code, since the previous 'fix' wasn't a fix.

joel-shor committed:
[tf.data] More debug code, since the previous 'fix' wasn't a fix.

joel-shor committed:
[tf.data] Fix BUILD file.

joel-shor committed:
[tf.data] Fix debug output.

joel-shor committed:
[tf.data] Fix debug output.

mrry commented on tensorflow/contrib/data/python/ops/interleave_ops.py:
```diff
@@ -240,9 +240,9 @@ def select_dataset(logits, seed):
-    logging.warn('dataset %i output_types: %s' % (i, dataset.output_types))
-    logging.warn('dataset %i output_shapes: %s' % (i, dataset.output_shapes))
+    logging.warn('dataset %i output_types: %s' % (i, str(dataset.output_types)))
+    logging.warn('dataset %i output_shapes: %s' % (i, str(dataset.output_shapes)))
```
(Not sure if you're still debugging this, but...) Can you please remove these `logging.warn()` calls before submitting?

joel-shor replied:
I am still debugging...I've just verified that the types and shapes of the datasets to `DirectedInterleaveDataset` are now the same between Windows and Linux, but I'm still getting the `kernel not registered` error. Any other suggestions for what might be causing this?

mrry commented on tensorflow/contrib/data/python/kernel_tests/resample_test.py:
```diff
@@ -109,5 +158,23 @@ def _remap_fn(_):
+    num_classes = len(init_dist)
+    # We don't need many samples to test a dirac-delta target distribution
+    num_samples = 1000
+    data_np = np.random.choice(num_classes, num_samples, p=init_dist)
```
This might be `np.int32` on Windows.

joel-shor committed:
[tf.data] Try fixing the Windows build by adding the directed interleave kernel to this cmake file.

joel-shor committed:
[tf.data] Remove debug code.

joel-shor commented:
@mrry I think the problem was that the interleave dataset op wasn't properly added to the cmake file. PTAL when you get a chance, once the tests are completed.

mrry commented:
Ah, good catch! I guess the `interleave_dataset_op_test.py` was [disabled due to an unrelated bug](https://github.com/tensorflow/tensorflow/blob/d0f5bc17560fc97bcc7de9164aa3b237a8d5221d/tensorflow/contrib/cmake/tf_tests.cmake#L285) and this left a hole in the test coverage.
